media-tv/mythtv

thebitpit opened the pull request:
Version bump to mythtv-30.0_p20190808 on fixes/30 branch bug #676952

Shortened verbose <use><flag> entries in metadata.xml and added myself as a proxy maintainer.
Updated to EAPI 7. 
Added user/group management per GLEP-81. 
This PR is my media-tv/mythtv_squash branch,  The messy details with many comments of how I got here are on my media-tv/mythtv branch PR  #12763.

Bug: https://bugs.gentoo.org/676952
Signed-off-by: Wilson Michaels <thebitpit@eartlink.net>

thebitpit commented:
Bug: https://bugs.gentoo.org/674216

juippis commented:
```
Calculating dependencies... done!
[ebuild  N     ] media-tv/mythtv-30.0_p20190808:0/30.0_p20190808::testworld  USE="alsa cdda cdr dvd exif wrapper x264 xml xvid -autostart -bluray -cec -ceton -debug -egl -fftw -hdhomerun -hdpvr -ieee1394 -ivtv -jack -java -lcd -libass -lirc -oss -perl -pulseaudio -python -raw -systemd -v4l -vaapi -vbox -vdpau -vpx -x265 -xmltv -zeroconf" INPUT_DEVICES="joystick" PYTHON_TARGETS="python2_7" 0 KiB

Total: 1 package (1 new), Size of downloads: 0 KiB

Would you like to merge these packages? [Yes/No] y

>>> Verifying ebuild manifests

>>> Emerging (1 of 1) media-tv/mythtv-30.0_p20190808::testworld
 * mythtv-30.0_p20190808.tar.gz BLAKE2B SHA512 size ;-) ...                         [ ok ]
/var/db/repos/testworld/media-tv/mythtv/mythtv-30.0_p20190808.ebuild: line 164: acct-user_src_install: command not found
>>> Unpacking source...
 * Unpacking mythtv-30.0_p20190808.tar.gz
>>> Source unpacked in /tmp/portage/media-tv/mythtv-30.0_p20190808/work
>>> Preparing source in /tmp/portage/media-tv/mythtv-30.0_p20190808/work/mythtv-30.0_p20190808/mythtv ...
>>> Source prepared.
>>> Configuring source in /tmp/portage/media-tv/mythtv-30.0_p20190808/work/mythtv-30.0_p20190808/mythtv ...
 * Running ./configure --prefix=/usr --libdir=/usr/lib64 --libdir-name=lib64 --mandir=/usr/share/man --cpu=host --enable-joystick-menu --disable-lirc --disable-libcec --disable-v4l2 --disable-ivtv --disable-hdpvr --disable-firewire --disable-hdhomerun --disable-vbox --disable-ceton --disable-audio-oss --enable-audio-alsa --disable-audio-jack --disable-audio-pulseoutput --disable-systemd_notify --disable-systemd_journal --without-bindings=perl,python --enable-libmp3lame --enable-libxvid --enable-libx264 --disable-libx265 --disable-libvpx --compile-type=release --enable-x11 --enable-xrandr --enable-xv --disable-vdpau --disable-vaapi --disable-libdns-sd --disable-libass --enable-symbol-visibility --enable-pic --disable-distcc --disable-ccache
ERROR: can not find libbluray.

If you think configure made a mistake, make sure that you are using the latest
version of MythTV from git.  If the latest version fails, report the problem to the
mythtv-dev@mythtv.org mailing list or IRC #mythtv on irc.freenode.net
Include the log file "config.ep" produced by configure as this will help
solve the problem.
>>> Source configured.
>>> Compiling source in /tmp/portage/media-tv/mythtv-30.0_p20190808/work/mythtv-30.0_p20190808/mythtv ...
make -j24 -l24 
Makefile:35: *** run configure to create config.mak.  Stop.
 * ERROR: media-tv/mythtv-30.0_p20190808::testworld failed (compile phase):
 *   emake failed
 * 
 * If you need support, post the output of `emerge --info '=media-tv/mythtv-30.0_p20190808::testworld'`,
 * the complete build log and the output of `emerge -pqv '=media-tv/mythtv-30.0_p20190808::testworld'`.
 * The complete build log is located at '/tmp/portage/media-tv/mythtv-30.0_p20190808/temp/build.log'.
 * The ebuild environment file is located at '/tmp/portage/media-tv/mythtv-30.0_p20190808/temp/environment'.
 * Working directory: '/tmp/portage/media-tv/mythtv-30.0_p20190808/work/mythtv-30.0_p20190808/mythtv'
 * S: '/tmp/portage/media-tv/mythtv-30.0_p20190808/work/mythtv-30.0_p20190808/mythtv'

>>> Failed to emerge media-tv/mythtv-30.0_p20190808
```

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,372 @@
+pkg_setup() {
+	python-single-r1_pkg_setup
+	# The acct-user/mythtv package creates/manages the user 'mythtv'
+	acct-user_src_install
```
This didn't work. Maybe just depend on acct-user/mythtv?

thebitpit replied:
Fixed emerge error.
Removed all acct_user_* calls.

juippis commented:
Well color me surprised, it actually compiled!

You should push acct-* ebuilds first, so the tree isn't broken at any state. What I mean here is that mythtv depends on those ebuilds in the first commit, while they aren't even in the tree yet. Then you should rebase and squash those fixes into mythv version bump. 

I'll do another review round for the ebuild, but something else for you to figure if you have time (and required for next version bump at least): 
```
 * QA Notice: Files built without respecting LDFLAGS have been detected
 *  Please include the following list of files in your report:
 * /usr/bin/mythavtest
 * /usr/bin/mythbackend
 * /usr/bin/mythccextractor
 * /usr/bin/mythcommflag
 * /usr/bin/mythexternrecorder
 * /usr/bin/mythfilerecorder
 * /usr/bin/mythfilldatabase
 * /usr/bin/mythfrontend.real
 * /usr/bin/mythjobqueue
 * /usr/bin/mythlcdserver
 * /usr/bin/mythmediaserver
 * /usr/bin/mythmetadatalookup
 * /usr/bin/mythpreviewgen
 * /usr/bin/mythreplex
 * /usr/bin/mythscreenwizard
 * /usr/bin/mythshutdown
 * /usr/bin/mythtranscode
 * /usr/bin/mythtv-setup
 * /usr/bin/mythutil
 * /usr/bin/mythwelcome
 * /usr/lib64/libmyth-30.so.30.0.0
 * /usr/lib64/libmythbase-30.so.30.0.0
 * /usr/lib64/libmythfreemheg-30.so.30.0.0
 * /usr/lib64/libmythmetadata-30.so.30.0.0
 * /usr/lib64/libmythprotoserver-30.so.30.0.0
 * /usr/lib64/libmythservicecontracts-30.so.30.0.0
 * /usr/lib64/libmythtv-30.so.30.0.0
 * /usr/lib64/libmythui-30.so.30.0.0
 * /usr/lib64/libmythupnp-30.so.30.0.0
 * /usr/lib64/mythtv/filters/libadjust.so
 * /usr/lib64/mythtv/filters/libbobdeint.so
 * /usr/lib64/mythtv/filters/libcrop.so
 * /usr/lib64/mythtv/filters/libdenoise3d.so
 * /usr/lib64/mythtv/filters/libfieldorder.so
 * /usr/lib64/mythtv/filters/libforce.so
 * /usr/lib64/mythtv/filters/libgreedyhdeint.so
 * /usr/lib64/mythtv/filters/libinvert.so
 * /usr/lib64/mythtv/filters/libivtc.so
 * /usr/lib64/mythtv/filters/libkerneldeint.so
 * /usr/lib64/mythtv/filters/liblinearblend.so
 * /usr/lib64/mythtv/filters/libonefield.so
 * /usr/lib64/mythtv/filters/libpostprocess.so
 * /usr/lib64/mythtv/filters/libquickdnr.so
 * /usr/lib64/mythtv/filters/libvflip.so
 * /usr/lib64/mythtv/filters/libyadif.so
```

You're doing good, I think it's close to being ready.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+
+MYTHTV_BRANCH="fixes/${P%.*}"
+
+DESCRIPTION="Homebrew PVR project"
```
(Just a thought but this could tell a little bit more)

thebitpit replied:
Improved.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+
+DESCRIPTION="Homebrew PVR project"
+HOMEPAGE="https://www.mythtv.org"
+SRC_URI="https://github.com/MythTV/mythtv/archive/${BACKPORTS}.tar.gz -> ${PF}.tar.gz"
```
`${P}` should be enough here, revbumps are meant to be used when ebuilds get enhancements, not new patch sets.

thebitpit replied:
Open Source DVR and media center hub

thebitpit replied:
OK I'll change it.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+"
+
+COMMON="
+	acct-user/mythtv
```
Why not add the group here as well?

thebitpit replied:
Primary group is a dependency of acct-user/mythtv ebuild. Mythtv requires a user named **mythtv** and does not expect a particular group. I use the **video** group on my machine and it works properly. _I have since learned this is a bad practice._

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+"
+DEPEND="${COMMON}
+	dev-lang/yasm
+	virtual/pkgconfig
```
This is defined in BDEPEND and can be removed from here.

thebitpit replied:
Done

juippis replied:
Still not done.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+
+Note that the systemd unit now restarts by default and logs
+to journald via the console at the notice verbosity.
+"
```
You could add a message about `pkg_config` phase and how to trigger it after installation.

thebitpit replied:
README.GENTOO-R1.ECLASS already shows this at first install only.

juippis replied:
I mean, you have this:
```
pkg_config() {
	echo "Creating mythtv MySQL user and mythconverg database if it does not"
	echo "already exist. You will be prompted for your MySQL root password."
	"${EROOT}"/usr/bin/mysql -u root -p < "${EROOT}"/usr/share/mythtv/database/mc.sql
}
```
But I never got prompted to do it, and wouldnt know to do it without reading this ebuild.

thebitpit replied:
Yes. I'll move the message.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+
+pkg_setup() {
+	python-single-r1_pkg_setup
+	# The acct-user/mythtv package creates/manages the user 'mythtv'
```
This comment is self-explanatory and can be removed.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+	local -a myconf
+
+	# Setup paths
+	myconf+=("--prefix=${EPREFIX}/usr")
```
When using array like this, you don't *need* `" "` inside it to my knowledge, but well, it didn't seem to break things either so use whatever style you want.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+	myconf+=("--mandir=${EPREFIX}/usr/share/man")
+
+	# CPU settings
+	for i in $(get-flag march) $(get-flag mcpu) $(get-flag mtune) ; do
```
Can you walk me through what this does exactly? Besides march should include mtune so you don't need to check both.

juippis replied:
> Can you walk me through what this does exactly? Besides march should include mtune so you don't need to check both.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+	has ccache ${FEATURES} || myconf+=("--disable-ccache")
+
+	einfo "Running ./configure ${myconf[@]}"
+	./configure \
```
Is there any way to make this configure script verbose? If not, add some note above that it might take some time...

juippis replied:
> Is there any way to make this configure script verbose? If not, add some note above that it might take some time...

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+		--extra-cxxflags="${CXXFLAGS}" \
+		--extra-ldflags="${LDFLAGS}" \
+		--qmake=$(qt5_get_bindir)/qmake \
+		${myconf[@]}
```
(if you disable `" "` inside array, it needs to be added around `"${myconf[@]}"`.

thebitpit replied:
So... is changing all
myconf+=("<whatever is there>")
to
myconf+=(<whatever is there>)
and using 
"${myconf[@]}"
a better style?

juippis replied:
I think it's less error-prone, but up to you really. It looks like this style also works.

thebitpit replied:
I'll change it. I like "less error-prone".

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+
+	insinto /usr/share/mythtv/contrib
+	# Ensure we don't install scripts needing the perl bindings (bug #516968)
+	use perl || find contrib/ -name '*.pl' -exec rm -f {} \;
```
Open this up and die on find properly.

thebitpit replied:
It is OK for **find** to fail if there are no files to remove.  Failure to remove a file with **rm -f** is OK too since it may have been passed a directory by find.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+	# Ensure we don't install scripts needing the perl bindings (bug #516968)
+	use perl || find contrib/ -name '*.pl' -exec rm -f {} \;
+	# Ensure we don't install scripts needing the python bindings (bug #516968)
+	use python || find contrib/ -name '*.py' -exec rm -f {} \;
```
Same here.

thebitpit replied:
Same as above.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+	python_fix_shebang "${ED}/usr/share/mythtv"
+
+	# Make shell & perl scripts executable
+	find "${ED}" -type f -name '*.sh' -o -type f -name '*.pl' | \
```
I mean, why `-exec` won't work here with find? I think you answered this but I lost it.

thebitpit replied:
Not done.  I tried, but I am not a good enough script writer to make it work.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+}
+
+pkg_info() {
+	if [[ -f "${EROOT}"/usr/bin/mythfrontend ]]; then
```
Same here, what's the point of this? I think I asked and you answered but let's get it recorded in this PR as well.

thebitpit replied:
It looks like an artifact from earlier ebuild that could build  mythbackend only. A mythfrontend may already be on the system. This warned the user they were out of sync. Removing it.  There is user demand for a mythbackend or mythfrontend only use flag option. It was noted in a comment that you suggested I remove and create a feature bug request.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,369 @@
+	fi
+
+	# Make Python files executable
+	find "${ED}/usr/share/mythtv" -type f -name '*.py' | while read file; do
```
This should probably be under `if use python; then`?

thebitpit replied:
Maybe. I going to leave it alone for now.

juippis replied:
You shouldn't. I didn't see any .py files under /usr/share in the source directory, or list of installed files even with python USE enabled. Seems like this whole section is useless, then?

thebitpit replied:
I found a lot of .py files under /usr/share/mythtv sub directories on my system. I moved mythtv to a backup directory and did a complete media-tv/mythtv emerge.  With python USE disabled, a lot of .py files are created in sub directories of /usr/share/mythtv .  They appear to be for support of some mythplugins. I am guessing that this is the reason mythtv must be _installed_ before mythplugins can be _compiled._  This code needs to be here until upstream fixes the packaging.

Then again this could just be my old system.

juippis replied:
> They appear to be for support of some mythplugins. I am guessing that this is the reason mythtv must be _installed_ before mythplugins can be _compiled._ This code needs to be here until upstream fixes the packaging.
> 

Hmm, indeed, I see them now.

thebitpit commented:
LDFLAGS:

LDFLAGS are already passed to the  "configure" script as extra LDFLAGS. Default setting and checking is done by "configure" for mythtv.  I thought the user had to set them in "make.conf".

juippis commented:
> LDFLAGS:
> 
> LDFLAGS are already passed to the "configure" script as extra LDFLAGS. Default setting and checking is done by "configure" for mythtv. I thought the user had to set them in "make.conf".

It means the program is not respecting user LDFLAGS, and you should try to patch it yourself or work with upstream so they fix it.

thebitpit commented:
I think I patched "configure" to respect LDFLAGS

I have never seen:

>  * QA Notice: Files built without respecting LDFLAGS have been detected
>  *  Please include the following list of files in your report:
>  * /usr/bin/mythavtest
>  * ...

How or where can I see it?
I use **repoman -dx full** to check and **emerge -v media-tv/mythtv** to test the ebuild.

juippis commented:
https://wiki.gentoo.org/wiki/Package_testing#make.conf
I use these settings when testing ebuilds, you can just append them before your emerge command.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,388 @@
+	bluray? ( xml )
+	cdr? ( cdda )
+"
+
```
It fails with `FEATURES="test" USE="test"` so you need to add `RESTRICT="test"` or fix them. Preferably fix them, but it can be saved for another bump. Anyway, try to study why they fail.

thebitpit replied:
I'll look into these later. They appear to involve audio QT interface type mismatch?
I added `RESTRICT="test"` and a short comment.

juippis commented:
Well, can't say I'm 100 % happy about this, but it works and doesn't have any more QA issues. So now if you squash your commits and add the acct-* commits to be before bumping mythtv, I'm happy.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,388 @@
+	systemd? ( sys-apps/systemd:= )
+	vaapi? ( x11-libs/libva:=[opengl] )
+	vdpau? ( x11-libs/libvdpau )
+	vpx? ( <media-libs/libvpx-1.7.0:= )
```
Oh also are you SURE about this? This is going to cause a lot of trouble for people with up-to-date systems.
Check with ldd and tools a-like please.

thebitpit replied:
I removed the restriction since the gentoo stable version in within the limitation.

thebitpit replied:
Stable vpx is `<media-libs/libvpx-1.7.0:=` so I removed the restriction.

> Check with ldd and tools a-like please.

I am not sure what to do with `ldd` so I ran `ldd -u` on the mythbackend binary.

`$ ldd -u /usr/bin/mythbackend`
Unused direct dependencies:
        /usr/lib64/libmythprotoserver-30.so.30
        /lib64/libpthread.so.0
        /usr/lib64/libQt5Gui.so.5
        /usr/lib64/libQt5Network.so.5
        /usr/lib64/libQt5Xml.so.5
        /usr/lib64/libQt5Sql.so.5
        /usr/lib64/libQt5Script.so.5
        /usr/lib/gcc/x86_64-pc-linux-gnu/8.3.0/libgcc_s.so.1

Removing some qt dependencies in the ebuild has no effect on the unused list.

thebitpit commented:
I left it in 3 pieces as it is easier for me to work will.  I will squash it when things are resolved.  
I keep finding and fixing things I did not know about.

juippis commented:
Some of my comments are still unanswered / unattended, also it really needs to be 3 commits for acct-* packages on their own and mythtv bump on its own. If you think you will mess everything up while trying, it can be done when merging but it's a bit more manual work for us (so we appreciate if you do it before).

thebitpit commented:
> Some of my comments are still unanswered / unattended,
I found one and agree the name should be `${P}.tar.gz`
> also it really needs to be 3 commits for acct-* packages on their own and mythtv bump on its own.
I will make it into 3 pull requests.

I grabbed media-libs/libhdhomerun because I use it with mythtv. I worked with Michał Górny and it is committed to gentoo repository. I got some ideas for small changes to make to this PR.

thebitpit commented:
I hope this is ready for merging. I have learned much about git and gentoo preparing this.

thebitpit commented:
On 9/15/19 12:11 PM, juippis wrote:
> Some of my comments are still unanswered / unattended, also it really 
> needs to be 3 commits for acct-* packages on their own and mythtv bump 
> on its own. If you think you will mess everything up while trying, it 
> can be done when merging but it's a bit more manual work for us (so we 
> appreciate if you do it before).
I believe I have addressed your comments. Please remind me of any I 
missed. The Pull request now contains 3 commits.

Please commit or inform me of any corrections anyone would like.

P.S. Yes I did "mess everything up". Then I figured out how to fix it, 
becoming more proficient with *git* along the way.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,406 @@
+	myconf+=($(use_enable pulseaudio audio-pulseoutput))
+
+	# Input Support
+	myconf+=($(use_enable lirc))
```
Just a note, you don't have to include `myconf+=` in every line here. You can just make it multi-line, like 
```
# Input Support
	myconf+=(
		$(use_enable lirc)
		$(use_enable input_devices_joystick joystick-menu)
		$(use_enable cec libcec)
		$(use_enable ieee1394 firewire)
	)
```
etc you get the idea, this applies to pretty much every section in this ebuild if you wish to make it cleaner.

thebitpit replied:
>     Can you walk me through what this does exactly? Besides march should include mtune so you don't need to check both.
I don't know why it is there. I'll look into it.

>  Just a note, you don't have to include myconf+= in every line here. You can just make it multi-line, like ...

I like your style much better.

thebitpit replied:
`Mythtv` uses the configure from `ffmpeg.` This bit of code is in the current `ffmpeg` ebuild. It tweaks some of the compiler optimizations.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,406 @@
+	# Make shell & perl scripts executable
+	find "${ED}" -type f -name '*.sh' -o -type f -name '*.pl' | \
+		while read file; do
+		chmod a+x "${file}"
```
`|| die`

thebitpit replied:
> Is there any way to make this configure script verbose?

I didn't find one. How about this:
`use debug  && einfo "Running ./configure ${myconf[@]}"`

Added `|| die `

thebitpit replied:
Reread your comment. Improved einfo message.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,406 @@
+}
+
+pkg_config() {
+	"${EROOT}"/usr/bin/mysql -u root -p < "${EROOT}"/usr/share/mythtv/database/mc.sql
```
I'm still not sure this does anything, didn't test this time though. Maybe https://devmanual.gentoo.org/ebuild-writing/functions/pkg_config/index.html is of some help?

thebitpit replied:
It only does something if the mythconverg database is missing.

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,406 @@
+	)
+"
+RDEPEND="${COMMON}
+	!media-tv/mythtv-bindings
```
Doesnt exist, not needed.

thebitpit replied:
Seems to be merged into mythtv

juippis commented on media-tv/mythtv/mythtv-30.0_p20190808.ebuild:
```diff
@@ -0,0 +1,406 @@
+"
+RDEPEND="${COMMON}
+	!media-tv/mythtv-bindings
+	!x11-themes/mythtv-themes
```
Doesnt exist, not needed.

thebitpit replied:
Seems to be merged into mythtv

juippis commented:
> I hope this is ready for merging. I have learned much about git and gentoo preparing this.

I think these are my last comments. This is a monstrous PR and task, but if it works after these are updated, I'm happy with it. Glad you've learned to use git tool efficiently, you'll become familiar with it :)

thebitpit commented:
>    virtual/pkgconfig
Still not done.

Right. it was half done. It is in BDEPEND, but also in DEPEND

thebitpit commented:
I attempted to address all your comments.

juippis commented:
Ok, I don't agree 100 % with the ebuild, but I did a few test runs with different USE flag combinations and it seems to work. Let's not hold it back any longer. I'll modify few things during merge, and don't be surprised if the next time you make a PR we ask you to change something that was already accepted. Let's get this into the tree, wait user feedback and bug reports. Thanks for your enormous work!
